Addition of some additional functionality for flash HIL

phil-levis opened the pull request:
This adds HIL interfaces for querying basic flash layout (where the kernel starts/stops, apps start) as well as locking flash pages.

Also renames FLASHCALW to FlashCalw to follow formatting conventions.

phil-levis commented:
Fixed formatting.

alevy commented on kernel/src/hil/flash.rs:
```diff
@@ -125,3 +127,36 @@ pub trait Client<F: Flash> {
     /// Flash erase complete.
     fn erase_complete(&self, error: Error);
 }
+
```
Comments

alevy commented on kernel/src/hil/flash.rs:
```diff
@@ -125,3 +127,36 @@ pub trait Client<F: Flash> {
 }
+
+pub trait FlashInfo {
+    fn flash_size(&self) -> u32;
```
For example, what are the units of the return value?

alevy replied:
Also, is this always num_pages * page_size? If so maybe there should be a default implementation.

alevy commented on kernel/src/hil/flash.rs:
```diff
@@ -125,3 +127,36 @@ pub trait Client<F: Flash> {
+    fn lock_unit_size(&self) -> u32;
+    /// Return 0 if flash does not support locking
+    fn pages_per_lock_unit(&self) -> u32;
+    fn page_to_lock_unit(&self, page: u32) -> u32;
```
What does this function do?

alevy commented on kernel/src/hil/flash.rs:
```diff
@@ -125,3 +127,36 @@ pub trait Client<F: Flash> {
+    fn page_to_lock_unit(&self, page: u32) -> u32;
+}
+
+pub trait FlashLocking {
```
This could probably use some deeper explanation, like what locking does, what a unit is, etc

brghena replied:
Also, do we expect Flash Locking to be a capability that all Flash hardware supports?

phil-levis replied:
Sorry, I thought this was obvious stuff, but obviously it isn't. Locking lets you make the flash unwriteable. I'll add comments explaining.

I'm hard pressed to think of a Cortex M that doesn't support locking in some form.  Every one I've seen does. It's pretty critical for a very wide range of applications. 

For example, I'd argue that now that we're bringing flash up, one of the first things the kernel should do is lock its own pages. We should therefore make sure that the kernel doesn't share a lock unit with applications.

alevy commented on kernel/src/hil/flash.rs:
```diff
@@ -125,3 +127,36 @@ pub trait Client<F: Flash> {
+}
+
+pub trait FlashLocking {
+    fn lock_unit(&self, page: u32);
```
Is the `page` argument in all these the start page of a unit?

phil-levis replied:
Should be unit

brghena commented on kernel/src/hil/flash.rs:
```diff
@@ -125,3 +127,36 @@ pub trait Client<F: Flash> {
+    fn unlock_units(&self, first: u32, last: u32);
+}
+
+pub trait FlashLayout {
```
Why is this part of the HIL interface for Flash? I could imagine external flash chips that run through this same interface but don't have kernel or apps on them. I'd expect these functions to be in the kernel somewhere.

phil-levis replied:
These are separate Traits. It could be that not every chip implements all of them. E.g., FlashLayout only makes sense for the on-chip flash. That being said, I think your point is well taken, the name is too general given what it means. I'd suggest "CodeLayout"? or "SoftwareLayout"?

brghena commented on chips/sam4l/src/flashcalw.rs:
```diff
@@ -946,6 +947,102 @@ impl hil::flash::Flash for FLASHCALW {
+    }
+
+    fn apps_end_address(&self) -> u32 {
+        self.get_flash_size() - 1
```
This isn't always correct (reinforcing my belief from another comment that these functions should be in the kernel). In Signpost, for instance, we reserve some top number of pages in flash for temporary application storage.

phil-levis replied:
If this isn't correct then it should be reflected in the linker map; we should add a symbol to kernel_layout.ld that reflects this. It shouldn't be a magic constant in software.

brghena replied:
For signpost, for example, we're using the `nonvolatile_storage_driver` to create a userspace-accessible region at the top of flash. It's not a part of the linker file, but rather the platform deciding to slice off a chunk of flash to use for a different purpose.
Example: https://github.com/lab11/signpost-software/blob/353afdbc1e0d41015d44707826fd270fd5a8afd4/kernel/boards/controller/src/main.rs#L600

My suggestion that the trait should be implemented by the kernel isn't exactly correct either though. In the example above, only that board actually knows the layout of flash space. Which I think makes sense in general for Tock. Individual platforms can make choices about where apps live and how much space to give them for both Flash and RAM. I'd argue that this is true for the kernel too, which could (although is not in practice) be placed anywhere by a platform.

Anyways, in general I just don't think this is a trait that Flash hardware is in any position to fulfill. You could create a simple capsule that lives on top of the FlashCalw, is fed this information at initialization time by the board, and could provide this interface though.

What's your intended usage for the "FlashLayout" commands?

phil-levis commented:
OK, I've moved FlashFormat into kernel/src/image, and added symbols to the kernel_layout.ld so that it's independent of the flash size and instead works off symbols defined in the linker scrupt.

Note that it still assumes that the entire application space is used for applications; if the last few pages are being used for something else, then this should somehow be incorporated into the linker script.

alevy commented:
I'm uncomfortable with how much information is being pushed to common linker script values. At the moment (i.e. in `master`), the only place linker script values are relied on is in the board crate, which makes sense because that's also where the linker script gets included.

The kernel crate has zero control or visibility into the linker script, however, this really ends up meaning that the kernel crate has a reverse dependency on the board crate (through an `extern`).

It's true that we have a common `kernel_layout.ld` linker script in the repo, but that's really a refactoring optimization given that we have several boards in the same repo. But, if someone makes a board outside the main repo, it's really up to them to write a linker script that fits their platform. In that case, this reverse dependency is problematic...

Perhaps making these functions part of the `Platform` trait would be better? That way the board author gets to choose where to take those values from. Then again, that working depends on how exactly you intend you use these, so it's tough to form an opinion on this change alone.

phil-levis commented:
I think it's reasonable to say that the kernel expects a certain set of symbols in order to operate properly. For example, it already depends on _etext, _srelocate, _erelocate, _szero, _ezero, _sapps, plus the boot sequence relies on the .app_memory link section. I think I'd flip this around and say that image should expose all of these through its HIL (or really, all of the information of the complete kernel RAM and ROM map). At some point, the kernel does depend very deeply on its image layout, although we do want to make sure that there's flexibility in it so it can support as wide a set of CortexMs as possible.

So what I'd suggest is:
  Update image to include the full set of symbols that the kernel expects/uses
    - I'd argue this should be start and end for every section.
  Update kernel_layout.ld to list this full set

Or are you considering cases when the kernel doesn't support any apps, or any storage, or?

alevy commented:
> For example, it already depends on _etext, _srelocate, _erelocate, _szero, _ezero, _sapps, plus the boot sequence relies on the .app_memory link section

The kernel crate does not. Those are _only_ referenced in the board-specific crates.

phil-levis commented:
Ah, now I understand what you're getting at. Makes sense. Is the goal generality or ease of developing a new platform? Because writing your own linker script is a pretty daunting task if you want to bring up a platform.

But it does seem a bit weird that we have several platforms, they all share the same linker script (admittedly in part due to the same MCU), but we'll be replicating the code for this.

brghena commented:
@phil-levis: a couple of thoughts about this.

1) What is your intended use case for the kernel start/end addresses and app start/end addresses? Understanding your plan for them would go a long way towards helping me understand if this is how we want to implement them for Tock.

2) I think, depending on that use case, that the kernel/app start/end address stuff should go into a separate pull request. Then we can discuss it separately from your FlashCalw and Flash trait changes and not have one side or the other holding the whole thing back. I know they made a lot of sense to combine at the start, but it feels like they've diverged at this point into two separate things.

3) As for this new approach, I think it's getting close to being what we want. Moving it to be part of the board makes sense. However, as currently implemented, every board's `main.rs` will have to change to add this code. Does your planned use apply to all boards so that they will have to implement this? Otherwise, if only some boards will need to provide this information, we could make it part of a new trait, rather than the `Platform` trait, so that boards could optionally implement it.

    If it is going to be required for each board to implement, it's annoying that it will just be a copy-paste job into each file. Each platform could make their own choices here and it's important to provide the ability for them to be able to set these however. But, maybe there's a happy middle-ground by doing something akin to the `load_processes` change from #511. There could be a default function in the kernel, which most platforms would call. This might be a bad idea though, if the goal is to keep the LD file symbols isolated to the board crate.
